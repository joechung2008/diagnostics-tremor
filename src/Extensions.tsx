import { byKey, isExtensionInfo, toNavLink } from "@/utils";
import { Button } from "@/components/Button";

const Extensions: React.FC<ExtensionsProps> = ({ extensions, onLinkClick }) => {
  const links = Object.values(extensions)
    .filter(isExtensionInfo)
    .map(toNavLink)
    .sort(byKey);

  return (
    <nav
      aria-label="Extensions"
      className="flex flex-col gap-2 max-h-max p-1 overflow-x-hidden"
    >
      {links.map((link) => (
        <Button
          key={link.key}
          variant="ghost"
          className="cursor-pointer min-h-[content] justify-start p-1 text-left w-full"
          onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
            onLinkClick?.(e, link);
          }}
        >
          {link.name}
        </Button>
      ))}
    </nav>
  );
};

export default Extensions;
