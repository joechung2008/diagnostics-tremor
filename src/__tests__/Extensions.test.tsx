import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import Extensions from "../Extensions";

describe("Extensions", () => {
  const mockOnLinkClick = vi.fn();

  const sampleExtensions = {
    ext1: {
      extensionName: "Extension A",
      config: { key: "value" },
    },
    ext2: {
      extensionName: "Extension B",
    },
    ext3: {
      lastError: {
        errorMessage: "Error occurred",
        time: "2023-01-01T00:00:00Z",
      },
    },
  };

  it("should render correctly with extensions", () => {
    const { container } = render(
      <Extensions extensions={sampleExtensions} onLinkClick={mockOnLinkClick} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should filter out ExtensionError and only show ExtensionInfo", () => {
    render(
      <Extensions extensions={sampleExtensions} onLinkClick={mockOnLinkClick} />
    );

    expect(screen.getByText("Extension A")).toBeInTheDocument();
    expect(screen.getByText("Extension B")).toBeInTheDocument();
    expect(screen.queryByText("Extension C")).not.toBeInTheDocument();
  });

  it("should sort extensions by key", () => {
    const unsortedExtensions = {
      z: { extensionName: "Z Extension" },
      a: { extensionName: "A Extension" },
      m: { extensionName: "M Extension" },
    };

    render(
      <Extensions
        extensions={unsortedExtensions}
        onLinkClick={mockOnLinkClick}
      />
    );

    const buttons = screen.getAllByRole("button");
    expect(buttons[0]).toHaveTextContent("A Extension");
    expect(buttons[1]).toHaveTextContent("M Extension");
    expect(buttons[2]).toHaveTextContent("Z Extension");
  });

  it("should call onLinkClick when button is clicked", () => {
    render(
      <Extensions extensions={sampleExtensions} onLinkClick={mockOnLinkClick} />
    );

    const button = screen.getByText("Extension A");
    fireEvent.click(button);

    expect(mockOnLinkClick).toHaveBeenCalledTimes(1);
    expect(mockOnLinkClick).toHaveBeenCalledWith(
      expect.any(Object), // MouseEvent
      expect.objectContaining({
        key: "Extension A",
        name: "Extension A",
        url: "",
      })
    );
  });

  it("should render empty when no ExtensionInfo items", () => {
    const errorOnlyExtensions = {
      ext1: {
        lastError: {
          errorMessage: "Error 1",
          time: "2023-01-01T00:00:00Z",
        },
      },
    };

    const { container } = render(
      <Extensions
        extensions={errorOnlyExtensions}
        onLinkClick={mockOnLinkClick}
      />
    );

    expect(container.firstChild).toHaveProperty("children", expect.any(Object));
    expect((container.firstChild as HTMLElement).children).toHaveLength(0);
  });

  it("should handle click without onLinkClick handler", () => {
    render(<Extensions extensions={sampleExtensions} />);

    const button = screen.getByText("Extension A");
    expect(() => fireEvent.click(button)).not.toThrow();
  });
});
