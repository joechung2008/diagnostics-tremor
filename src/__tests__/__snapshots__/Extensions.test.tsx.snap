// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Extensions > should render correctly with extensions 1`] = `
<nav
  aria-label="Extensions"
  class="flex flex-col gap-2 max-h-max p-1 overflow-x-hidden"
>
  <button
    class="relative inline-flex items-center whitespace-nowrap rounded-md border text-sm font-medium transition-all duration-100 ease-in-out disabled:pointer-events-none disabled:shadow-none outline-offset-2 outline-0 focus-visible:outline-2 outline-blue-500 dark:outline-blue-500 shadow-none border-transparent text-gray-900 dark:text-gray-50 bg-transparent hover:bg-gray-100 dark:hover:bg-gray-800/80 disabled:text-gray-400 dark:disabled:text-gray-600 cursor-pointer min-h-[content] justify-start p-1 text-left w-full"
    tremor-id="tremor-raw"
  >
    Extension A
  </button>
  <button
    class="relative inline-flex items-center whitespace-nowrap rounded-md border text-sm font-medium transition-all duration-100 ease-in-out disabled:pointer-events-none disabled:shadow-none outline-offset-2 outline-0 focus-visible:outline-2 outline-blue-500 dark:outline-blue-500 shadow-none border-transparent text-gray-900 dark:text-gray-50 bg-transparent hover:bg-gray-100 dark:hover:bg-gray-800/80 disabled:text-gray-400 dark:disabled:text-gray-600 cursor-pointer min-h-[content] justify-start p-1 text-left w-full"
    tremor-id="tremor-raw"
  >
    Extension B
  </button>
</nav>
`;
